### PR TITLE
User history stream [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by phamann**
Opening this now for review, but won't merge until have solved the Heroku Java problem.

### This PR:
- Adds a new worker process `server/jobs/user-history.js` which reads data from Spoor's Amazon Kinesis stream, using the KCL client.
- If a given event matches our criteria passes to `server/models/History.js` to save:
    - Decorate event with user session info, specifically the `uuid`
    - Decorate event with the tag/concept IDs for a give path
    - Save to DynamoDB table with timestamp and uuid.

### Todo:
- [ ] Push success/failure metrics somewhere, Graphite/Splunk?
- [ ] Work out how to deploy multipack java/node (maybe just curl a java bin?)
- [ ] Add support for www.ft.com events
- [ ] Expose Dynamo get/query as api endpoint for Ilyushin to consume.
- [ ] Tests :wink: 

### Scaling questions
- [ ] Do we need to consider a different `ProvisionedThroughput` limit on the DB table as it will be written to quite frequently?
- [ ] Should we have another SQS queue in-between the job process and the database model saving?

/cc @wheresrhys @adgad @commuterjoy

### CODE REVIEW

File: scripts/dynamodb/007-create-history-table.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/*jshint node:true*/
+
+'use strict';
+
+var AWS = require('aws-sdk');
+
+AWS.config.update({
```
Comment (wheresrhys) [inline scripts/dynamodb/007-create-history-table.js]: there's a `./lib/aws-connect` module for this.

File: server/jobs/user-history.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+var kcl = require('aws-kcl');
+var history = require('../models/History');
+var articleRegEx = '^/([a-f0-9]+)-([a-f0-9]+)-([a-f0-9]+)-([a-f0-9]+)-([a-f0-9]+)';
+var streamRegEx = '^/stream/(brandId|peopleId|sectionsId|organisationsId|regionsId|topicsId|authorsId|subjectsId|specialReportsId|icbId)/(.*)';
+
+function isStreamPath(path) {
+	return new RegExp(streamRegEx).test(path);
```
Comment (wheresrhys) [inline server/jobs/user-history.js]: why generate new regex each time?  Why not just store articleRegEx and streamRegEx as actual regexes?

File: server/jobs/user-history.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		},
+
+		processRecords: function(processRecordsInput, completeCallback) {
+			if (!processRecordsInput || !processRecordsInput.records) {
+				completeCallback();
+				return;
+			}
+
+			var records = processRecordsInput.records;
+			var record, sequenceNumber, partitionKey, data, rawData;
```
Comment (wheresrhys) [inline server/jobs/user-history.js]: Sorry for being totally anal, but theres' a [vague] convention in next of declaring each variable using `var` on a new line.

File: server/jobs/user-history.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+						session: data.annotations.session,
+						path: data.annotations.referer.pathname,
+						type: (isStreamPath(data.annotations.referer.pathname)) ? 'stream' : 'article'
+					});
+				}
+			}
+
+			// Checkpoint last sequence number.
+			processRecordsInput.checkpointer.checkpoint(sequenceNumber, function(err, sn) {
+				// TODO Error handling logic, do we need any?
```
Comment (wheresrhys) [inline server/jobs/user-history.js]: could wrap the whole thing in a promise to be safe and think about doing something with the error later?

File: server/lib/api.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+require('isomorphic-fetch');
```
Comment (wheresrhys) [inline server/lib/api.js]: Is this called within a job or the app? In general only include isomorphic fetch in top level job files - in apps next-express adds it as a global already.

In fact, we need to think about replicating https://github.com/Financial-Times/next-express/blob/master/main.js#L129 when fetch is used in a worker process

File: server/lib/api.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+var api = require('next-ft-api-client');
+
+function getArticleContent(path) {
+	return api.contentLegacy({
+		uuid: path.replace('/', ''),
+		useElasticSearch: false
```
Comment (wheresrhys) [inline server/lib/api.js]: not sure what the state of play is with elastic search, but possibly we should have true here?

File: server/models/History.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
+'use strict';
```
Comment (wheresrhys) [inline server/models/History.js]: to minimise merge problems with my PR I'm gonna add a file `server/models/tables/history.js` in master. Could you move the contents of this file into that (unless you think my restructuring of this directory is a bad idea, in which case let me know)

**Comment by wheresrhys** [comment]
Aside from comments :+1:

File: server/jobs/user-history.js (hunk 2 above)
Comment (commuterjoy) [inline server/jobs/user-history.js]: Safer to stick a dollar at the end of these?

**Comment by commuterjoy** [comment]
Good work.

Speak to Gathercole about multi buildpacks?

File: server/jobs/user-history.js (hunk 4 above)
Comment (phamann) [inline server/jobs/user-history.js]: Sadly, as this the interface the AWS client is expecting we cannot wrap in a promise. The question is do we care that fetching an event failed, as its a large stream I'm not too concerned about processing every single event. But would care if we failed to write an event to Dynamo for instance.

File: scripts/dynamodb/007-create-history-table.js (hunk 1 above)
Comment (phamann) [inline scripts/dynamodb/007-create-history-table.js]: Am going to leave this, as its just the create table script and follows same convention as the others.

File: server/jobs/user-history.js (hunk 4 above)
Comment (adgad) [inline server/jobs/user-history.js]: Might be worth logging to graphite (or splunk?). Individual fails not so important but we should have some visibility of something going badly wrong e.g. no events processed.

**Comment by phamann** [comment]
Closing this as we have extracted into it's own application:

https://github.com/Financial-Times/next-spoor-ingest

**Pull request closed without merging**